layers/+lang/latex/packages.el

```diff
@@ -196,6 +197,22 @@
 (defun latex/post-init-flyspell ()
   (spell-checking/add-flyspell-hook 'LaTeX-mode-hook))
 
+(defun latex/init-evil-tex ()
+  (use-package evil-tex
+    :defer t
+    :init
+    (progn
+      (add-hook 'LaTeX-mode-hook #'evil-tex-mode)
+      (setq evil-tex-toggle-override-m nil)
+      (setq evil-tex-toggle-override-t nil))
+    :config
+    (progn
+      (spacemacs/set-leader-keys-for-major-mode 'latex-mode
+        "e" evil-tex-toggle-map)
+      (spacemacs/declare-prefix-for-mode 'latex-mode "mm" "magnificent toggles")
```

Comment:
I think this needs to be added to :init block otherwise it won't show up until it's loaded.

Reply:
I think thanks to the magic of auto-loading this "works" (since switching to a latex buffer invokes evil-tex via the hook (which was placed there in init and so then this config block runs). I could move it into init, but not sure if evil-tex-toggle-map is defined prior to evil-tex loading (though I can check)

Reply:
Yea, I can't move it to :init because evil--tex-toggle-map is not defined when that block runs. I suppose I could move the prefix declaration there, but seems better to keep them packaged together.

Reply:
what I meant is to move declare-prefix alone. (not the actual key bindings)

Reply:
Is "declare-prefix" the "idiomatic" spacemacs way to give pretty labels to prefix keymaps in which-key? I just followed the pattern I saw elsewhere in the file.

Reply:
spacemacs/declare-prefix-for-mode is the idiomatic way for that in a major/minor mode specific keymap.

Reply:
Moving the declaration into :init doesn't work either. It doesn't throw an error, but when it actually gets bound it overwrites the string and just shows up as +prefix in which-key. Everywhere else I've seen it used the prefix key gets mapped first and then declare-prefix follows (presumably for this reason). I'll go back to wrapping them in progn in :config so that I can do them in the proper order.

Reply:
> It doesn't throw an error, but when it actually gets bound it overwrites the string and just shows up as +prefix in which-key. 

well this is a known problem when the same prefix is used again...

Reply:
Sorry, I don't follow. If I want to put a given keymap behind a key *and* give it a "friendly" name in which-key, what am I supposed to do?

Reply:
it's not documented.

```
(spacemacs/declare-prefix-for-mode 'foo-mode
  "<prefix-keys>" '("description . map))
```

something like this adds both description and key bindings.
(you can check the documentation for which-key-add-keymap-based-replacements)

Reply:
No luck. It complains about evil-tex-toggle-map failing commandp

Reply:
check documentation of which-key-add-keymap-based-replacements

I may be mistaken on the calling convention.

Reply:
From the definition of spacemacs/declare-prefix-for-mode and what it calls, it seems like it doesn't handle any mapping, just changing descriptions (it calls which-key-add-keymap-based-replacements), so it doesn't seem to me like there is an obvious "one-shot" way of doing the mapping and the relabeling at once, and instead one needs to first do the mapping, and then do the relabeling to avoid the former overwriting the latter.

I do see a note about preferring the "cons" style for replacement being preferred by which-key, but when the command is a prefix, then command is nil and it's a trivial cons.

Reply:
which-key-add-keymap-based-replacements should add both description and bindings.

Reply:
> but when the command is a prefix, then command is nil and it's a trivial cons

the behavior is actually subtly different in terms of font face... that's not documented.

Reply:
> which-key-add-keymap-based-replacements should add both description and bindings.

I don't think so. I tried evaluating the example from the which-key documentation:
```
(which-key-add-keymap-based-replacements global-map
  "C-x w" '("Save as" . write-file))
```

yet `C-x w` remains bound to window selection for me

Reply:
see this

https://github.com/justbur/emacs-which-key/blob/1217db8c6356659e67b35dedd9f5f260c06f6e99/which-key.el#L923

Reply:
Fair enough. I've gone a bit further down this rabbit hole and just can't figure out a way to get this to work. Consider the following elisp, inspired by [this section of the `which-key` documentation](https://github.com/justbur/emacs-which-key#keymap-based-replacement)

```
(define-key global-map (kbd "C-t") '("My keymap" . evil-tex-toggle-map))
(define-key global-map (kbd "C-t") '("My keymap" . (evil-tex-toggle-map)))
(define-key global-map (kbd "C-t") '("My keymap" . 'evil-tex-toggle-map))
(define-key global-map (kbd "C-t") evil-tex-toggle-map)
```

The last line works as I'd expect and causes `C-t` to be a prefix key for `evil-tex-toggle-map` (when `C-t` isn't intercepted by a higher priority keymap), but of course doesn't set the description "My keymap".

The first three lines all run but fail to effect a binding. If I try to fire `C-t`, they all fail with `Wrong type argument: commandp, x` where `x` is the cdr of the cons that I used.

I looked at the [documentation for `define-key`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Changing-Key-Bindings.html#index-define_002dkey), where the argument in question is `binding`, which referred me to the documentation for [key lookup](https://www.gnu.org/software/emacs/manual/html_node/elisp/Key-Lookup.html). Assuming that documentation is correct, it seems that if `binding` is a list, then it's `car` needs to either be the symbol `keymap` or `lambda`; there's nothing about using the car for a string and putting the function in the cdr.

Reply:
It seems I am not [the only one confused by this](https://github.com/justbur/emacs-which-key/issues/253)/

Here is the approach that works
```
(define-key global-map (kbd "C-t") (cons "My keymap" evil-tex-toggle-map))
```

I'd argue that the documentation for `which-key` is a bit confusing, but maybe my elisp-skills are just too poor to understand what's going on.

Reply:
and returning to my specific example, here is a command which in the :config block remaps the key and sets the description all in one shot, as you suggested

```
(spacemacs/declare-prefix-for-mode 'latex-mode "mq"
        (cons "evil-tex toggles" evil-tex-toggle-map))
```